### PR TITLE
Use endpoints to fetch field values for card-based parameters

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/26898

How to test:
- New -> Native
- `SELECT * FROM PRODUCTS WHERE {{tag}}`
- Field Filter -> Products -> Category
- Dropdown -> Edit -> Custom list -> Add `Gadget` and `Widget`
- **Save the question**
- Click on the filter. It should now use a different BE endpoint and display only `Gadget` and `Widget`

Please note that this PR target the integration branch and not master.

Screenshot 2023-01-20 at 16 48 16

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/27727)

### CODE REVIEW

File: frontend/src/metabase-lib/parameters/utils/parameter-source.ts [hunk 1]
```diff
@@ -26,15 +27,10 @@ export const canUseCustomSource = (parameter: Parameter) => {
...
-  switch (type) {
-    case "string":
-    case "location":
-      return subType === "=";
-    case "category":
-      return true;
-    default:
-      return false;
-  }
+  return (
```
Comment (ranquild) [inline frontend/src/metabase-lib/parameters/utils/parameter-source.ts:30]: This reverts the change made by a previous PR

File: frontend/src/metabase/components/FieldValuesWidget.jsx [hunk 2]
```diff
@@ -46,11 +47,14 @@ const optionsMessagePropTypes = {
   message: PropTypes.string.isRequired,
 };
 
-const mapDispatchToProps = {
-  addRemappings,
-  fetchFieldValues,
-  fetchDashboardParameterValues,
-};
+const mapDispatchToProps = dispatch => ({
```
Comment (ranquild) [inline frontend/src/metabase/components/FieldValuesWidget.jsx:50]: Needed to resolve a dependency cycle between imports

File: frontend/src/metabase/components/FieldValuesWidget.jsx [hunk 3]
```diff
@@ -452,9 +471,11 @@ OptionsMessage.propTypes = optionsMessagePropTypes;
 
 export default connect(mapStateToProps, mapDispatchToProps)(FieldValuesWidget);
 
-// if [dashboard] parameter ID is specified use the fancy new Chain Filter API endpoints to fetch parameter values.
-// Otherwise (e.g. for Cards) fall back to the old field/:id/values endpoint
-function usesChainFilterEndpoints(dashboard) {
+function canUseQuestionEndpoints(question) {
+  return question?.isSaved();
```
Comment (ranquild) [inline frontend/src/metabase/components/FieldValuesWidget.jsx:475]: Currently new filtering works only with saved cards

File: frontend/src/metabase/query_builder/actions/parameters.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+interface FetchParameterValuesOpts {
+  question: Question;
+  parameter: Parameter;
+  query?: string;
+}
+
+export const FETCH_QUESTION_PARAMETER_VALUES =
+  "metabase/qb/FETCH_QUESTION_PARAMETER_VALUES";
+
+export const fetchQuestionParameterValues = createThunkAction(
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/actions/parameters.ts:17]: Here I'm replicating the same logic used for caching parameter values in dashboards.

File: frontend/test/__support__/e2e/commands/api/dashboard.js [hunk 5]
```diff
@@ -1,9 +1,23 @@
...
+    embedding_params = {},
+    ...dashboardDetails
+  } = {}) => {
     cy.log(`Create a dashboard: ${name}`);
 
     // For all the possible keys, refer to `src/metabase/api/dashboard.clj`
-    return cy.request("POST", "/api/dashboard", { name, ...dashboardDetails });
+    cy.request("POST", "/api/dashboard", { name, ...dashboardDetails }).then(
+      ({ body }) => {
+        if (enable_embedding) {
```
Comment (ranquild) [inline frontend/test/__support__/e2e/commands/api/dashboard.js:14]: A similar approach used in `createQuestion`. For flags that aren't supported in `POST` endpoints, issue a `PUT` immediately.

File: frontend/src/metabase/components/FieldValuesWidget.jsx (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase/components/FieldValuesWidget.jsx:475]: Next PR, for the context https://github.com/metabase/metabase/pull/27799

File: frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js [hunk 6]
```diff
@@ -243,3 +243,19 @@ export function saveQuestion(
...
+  );
+}
+
+export function visitPublicDashboard(id) {
+  cy.request("POST", `/api/dashboard/${id}/public_link`).then(
+    ({ body: { uuid } }) => {
+      cy.visit(`/public/dashboard/${uuid}`);
+    },
+  );
+}
```
Comment (kulyk) [inline frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js:261]: Does it make sense to intercept and wait for `GET /api/public/card/:uuid` (and dashboard endpoint too)?

File: frontend/src/metabase/public/components/EmbedFrame.jsx [hunk 7]
```diff
@@ -78,7 +80,8 @@ class EmbedFrame extends Component {
                 <div className="flex">
                   <SyncedParametersList
                     className="mt1"
-                    dashboard={this.props.dashboard}
+                    question={question}
+                    dashboard={dashboard}
```
Comment (kulyk) [inline frontend/src/metabase/public/components/EmbedFrame.jsx:84]: Changes to `EmbedFrame` will most likely result in conflicts with `master` after #27746. Probably it'd make sense to sync the base branch and this branch with the latest `master`

File: frontend/src/metabase/parameters/components/widgets/ParameterFieldWidget/ParameterFieldWidget.jsx [hunk 8]
```diff
@@ -74,6 +76,7 @@ export default function ParameterFieldWidget({
               value={value}
               parameter={parameter}
               parameters={parameters}
+              question={question}
```
Comment (kulyk) [inline frontend/src/metabase/parameters/components/widgets/ParameterFieldWidget/ParameterFieldWidget.jsx:79]: Woah, what a path it had to take: `PublicQuestion` > `EmbedFrame` > `SyncedParametersList` > `ParametersList` > `ParameterWidget` > `ParameterValueWidget` > `ParameterFieldWidget` > `FieldValuesWidget`

I understand that the answer is probably "no", but is there something we can do to shorten this path?

File: frontend/src/metabase/query_builder/actions/parameters.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+export const fetchQuestionParameterValues =
+  ({ question, parameter, query }: FetchParameterValuesOpts) =>
+  async () => {
+    const apiArgs = { cardId: question.id(), paramId: parameter.id, query };
+
+    const { values, has_more_values } = query
+      ? await CardApi.parameterSearch(apiArgs)
+      : await CardApi.parameterValues(apiArgs);
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/actions/parameters.ts:18]: While looking at QB code, I expect `query` to be an instance of a `StructuredQuery` or a `NativeQuery`. It doesn't seem to be the case here though. Is there a different name we can use? I literally was trying to figure out why the way we choose an endpoint depends on the presence of a question query 😅

**Review by kulyk — APPROVED** [review]

File: frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js (hunk 6 above)
Comment (ranquild) [reply frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js]: Great catch!

**Pull request merged**